doozerlib/util.py

```diff
@@ -96,6 +98,34 @@ def convert_remote_git_to_https(source):
     return re.sub(string=url, pattern=r'\.git$', repl='').rstrip('/')
 
 
+def setup_and_fetch_public_upstream_source(public_source_url: str, public_upstream_branch: str, source_dir: str, set_env: Dict[str, str] = None):
+    cmd = ["git", "-C", source_dir, "remote", "get-url", "public_upstream"]
+    rc, out, err = exectools.cmd_gather(cmd)
+    if rc != 0:
+        if "No such remote" in err:
+            exectools.cmd_assert(["git", "-C", source_dir, "remote", "add", "--", "public_upstream", public_source_url])
+        else:
+            raise IOError(f"Error running {cmd}. See debug log.")
+    elif out.strip() != public_source_url:
+        exectools.cmd_assert(["git", "-C", source_dir, "remote", "set-url", "--", "public_upstream", public_source_url])
+    exectools.cmd_assert(["git", "-C", source_dir, "fetch", "--", "public_upstream", public_upstream_branch], retries=3, set_env=set_env)
+
+
+def is_commit_in_public_upstream(revision: str, public_upstream_branch: str, source_dir: str):
+    # rc, out, err = exectools.cmd_gather(["git", "-C", source_dir, "branch", "--remotes", "--list", "public_upstream/*", "--contains", revision])
+    # if rc == 0:
+    #     return out.strip() != ""
+    # if "no such commit" in err:
+    #     return False
+    cmd = ["git", "merge-base", "--is-ancestor", "public_upstream/" + public_upstream_branch, "--", revision]
+    rc, out, err = exectools.cmd_gather(cmd)
+    if rc == 0:
+        return True
+    if err == "" or "Not a valid commit name" in err:
```

Comment:
Same caution on using stderr messages. Does `--contains` not work as expected?

Reply:
`--contains` works, but it doesn't set exit code. It seems to me `merge-base --is-ancestor` is more reliable to distinguish `commit doesn't exist in a branch` and `general command error`.

Reply:
What I want to achieve here is just to distinguish `commit doesn't exist` from other command execution errors. Another choice is using a library like [pygit2](https://github.com/libgit2/pygit2) to do that.

Reply:
@jupierce The doc https://git-scm.com/docs/git-merge-base#Documentation/git-merge-base.txt---is-ancestor said the exit code is 1 if the commit is not in the branch, while for other errors the exit code is not 1. It is clearly documented so I think we can use the exit code.

Reply:
If `--contains` runs and exits with zero, you simply trust the content of stdout. If the stdout does not include the upstream branch, you know to set private_fix.